curl: enable http2 support

TSAnsari opened the pull request:
avs-device-sdk relies on http2 connection to communicate
with the Alexa voice service. This commit uses nghttp2
library to enable http2 support for curl.

Signed-off-by: Tariq Ansari <tansari@luxoft.com>

sashko commented on recipes-support/curl/curl_%.bbappend:
```diff
@@ -0,0 +1,3 @@
+PACKAGECONFIG_append = " nghttp2 "
+PACKAGECONFIG_remove_class-native = " nghttp2 "
+PACKAGECONFIG_remove_class-nativesdk = " nghttp2 "
```
what do we need this for?

TSAnsari replied:
To configure curl to enable nghttp2

sashko replied:
I've meant the `-nativesdk` part

TSAnsari replied:
so when I append nghttp2 to PACKAGECONFIG, it also appends to class-native and class-nativesdk. So I have to explicitly remove it.

sashko replied:
but why?

martin-ejdestig replied:
Also, it is a bit unclear to me what uses curl. I assume it is a library in the avs-device-sdk ? If so, is there no way to say "depend on nghttp2 being enabled in curl" in the avs-device-sdk itself so a bbappend is not needed? Or maybe that is not how it is done. @sashko ?

sashko replied:
yes, I believe we should add `nghttp2` packageconfig only when the sample app is built.

martin-ejdestig replied:
Hum, searching through a soruces/ dir of a Pelux build I can not see other recipes modifying PACKAGECONFIG for other packages...

@DunderRoffe Do you know how to do this (enable http2 for curl in recipe of Alexa SDK)... if it is something that should be done?

TSAnsari replied:
> yes, I believe we should add `nghttp2` packageconfig only when the sample app is built.

Its not the sample application. Its the avs-device-sdk that requires curl with http2 enabled.

TSAnsari replied:
> Also, it is a bit unclear to me what uses curl. I assume it is a library in the avs-device-sdk ? If so, is there no way to say "depend on nghttp2 being enabled in curl" in the avs-device-sdk itself so a bbappend is not needed? Or maybe that is not how it is done. @sashko ?

Yes it is avs-device-sdk that requires it and it explicitly checks for curl-http2 support during initialization. 
So incase we somehow move the curl-http2 PACKAGECONFIQ to avs-device-sdk. Then in the future we want support for curl-http2 for e.g. in SOTA service and exclude avs-sdk, then we need to add the curl-http2 support in sota recipe also.
Why not enable http2 in curl as default, since we see future need for it?

DunderRoffe replied:
Take a look at https://code.qt.io/cgit/yocto/meta-boot2qt.git/tree/meta-boot2qt/recipes-qt/automotive/qtivi_git.bb
PACKAGECONFIG is split up into several groups which makes it clear what feature they enable. Then you can specify which groups shuld be enabled

TSAnsari replied:
> but why?

So what I understood is that when we append nghttp2 to PACKAGECONFIG it is also appended to all the classes(i.e. native, nativesdk etc) for which the recipe has been extended. Since for the native and nativesdk of curl we do not want to enable http2, so we remove it.

TSAnsari commented:
> it's nghttp2, not http; curl already supports http out of the box
> 
> also, append itself is for curl, not Alexa

I never said http. In the commit message it is http2.

Yes you are correct. I will replace Alexa with curl.

sashko commented:
> > it's nghttp2, not http; curl already supports http out of the box
> > also, append itself is for curl, not Alexa
> 
> I never said http. In the commit message it is http2.
> 
> Yes you are correct. I will replace Alexa with curl.

but this `nghttp2`, specific library you add support for.

TSAnsari commented:
So after some discussion above and face to face discussions, it was concluded that we don't have a way of enabling curl-http2 configuration from another recipe (i.e. avs-device-sdk). So we keep the changes as it is.

sashko commented:
@TSAnsari why was this merged without addressing comments given during the code-review?

TSAnsari commented:
> @TSAnsari why was this merged without addressing comments given during the code-review?

Because all the review comments were addressed.

sashko commented:
> > @TSAnsari why was this merged without addressing comments given during the code-review?
> 
> Because all the review comments were addressed.

> ok, just change it to nghttp2, coz this is what you're adding

TSAnsari commented:
> > > @TSAnsari why was this merged without addressing comments given during the code-review?
> > 
> > 
> > Because all the review comments were addressed.
> 
> > ok, just change it to nghttp2, coz this is what you're adding

Its mentioned in the commit message

sashko commented:
> > > > @TSAnsari why was this merged without addressing comments given during the code-review?
> > > 
> > > 
> > > Because all the review comments were addressed.
> > 
> > 
> > > ok, just change it to nghttp2, coz this is what you're adding
> 
> Its mentioned in the commit message

This reminds me Tinder, where women set their age to 29, but in the bio say "I am actually 36".
